# Проект по SQL

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследование таблиц</a></span></li><li><span><a href="#Задание-1" data-toc-modified-id="Задание-1-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Задание 1</a></span></li><li><span><a href="#Задание-2" data-toc-modified-id="Задание-2-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задание 2</a></span></li><li><span><a href="#Задание-3" data-toc-modified-id="Задание-3-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задание 3</a></span></li><li><span><a href="#Задание-4" data-toc-modified-id="Задание-4-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задание 4</a></span></li><li><span><a href="#Задание-5" data-toc-modified-id="Задание-5-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задание 5</a></span></li></ul></div>

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Задача — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**

Таблица `books` содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors` содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers` содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

Таблица `ratings` содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews` содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Задания:**

- Посчитать, сколько книг вышло после 1 января 2000 года.
- Для каждой книги посчитать количество обзоров и среднюю оценку.
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Исследование таблиц

Подготовим подключение к базе:

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Чтобы выполнить SQL-запрос, используем Pandas:

In [3]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 
    display(request.head())
    print(request.info())
    print(f'Таблица - "{table}"')
    print('______________________________________________________________________________________________')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица - "books"
______________________________________________________________________________________________


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица - "authors"
______________________________________________________________________________________________


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица - "publishers"
______________________________________________________________________________________________


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица - "ratings"
______________________________________________________________________________________________


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица - "reviews"
______________________________________________________________________________________________


Для удобства дальнейшего вывода результатов запросов на экран используем функцию:

In [4]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

Пропусков нет, но нужно поменять тип данных в столбце `publication_date` таблицы `books`. Сделаем это после того как запишем таблицы в двтвфреймы:

In [5]:
select_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(select_books, con = engine)

select_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(select_reviews, con = engine)

select_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(select_ratings, con = engine)

select_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(select_publishers, con = engine)

select_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(select_authors, con = engine)

Приведем данные к нужному типу:

In [6]:
books['publication_date'] = books['publication_date'].astype('datetime64')

## Задание 1

Посчитаем, сколько книг вышло после 1 января 2000 года.

In [7]:
books_count = '''
SELECT COUNT(book_id) AS books_count
FROM books
WHERE publication_date >= '2000-01-01'
'''
select(books_count)

books_count
0          821

**Вывод:**

После 1 января 2000 года вышла 821 книга.

## Задание 2

Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [8]:
ratings_reviews = '''
SELECT 
    books.title AS book_title,
    books.book_id AS id,
    AVG(ratings.rating) AS avg_rating,
    COUNT(DISTINCT reviews.text) AS reviews_count
FROM
    books
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY
     book_title,
     id
ORDER BY
     reviews_count DESC
     -- reviews_count
     -- avg_rating DESC
     -- avg_rating DESC
-- LIMIT 20;
'''
select(ratings_reviews)

book_title   id  avg_rating  \
0                              Twilight (Twilight  #1)  948    3.662500   
1                                  Water for Elephants  963    3.977273   
2                                     The Glass Castle  734    4.206897   
3    Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
4    The Curious Incident of the Dog in the Night-Time  695    4.081081   
..                                                 ...  ...         ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel   83    3.666667   
996                            The Natural Way to Draw  808    3.000000   
997   The Cat in the Hat and Other Dr. Seuss Favorites  672    5.000000   
998                          Essential Tales and Poems  221    4.000000   
999  Disney's Beauty and the Beast (A Little Golden...  191    4.000000   

     reviews_count  
0                7  
1                6  
2                6  
3                6  
4                6  
..             ...  
995              0  
996              0  
997              0  
998              0  
999              0  

[1000 rows x 4 columns]

**Вывод:**

Максимальный рейтинг - 5, минимальный - 1,5. Максимальное количество обзоров у книги Twilight (Twilight #1). Есть 6 книг, которые остались без пользовательских обзоров. Достаточно много книг с оценкой 5, но при этом небольшим количеством обзоров.

## Задание 3

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.

In [9]:
published_max = '''
SELECT
     publishers.publisher AS publisher_name,
     COUNT(books.book_id) AS books_published
FROM
    publishers
LEFT JOIN books ON books.publisher_id = publishers.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    publisher_name
ORDER BY 
    books_published DESC
LIMIT 1;
'''
select(published_max)

publisher_name  books_published
0  Penguin Books               42

**Вывод:**

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books.

## Задание 4

Определим автора с самой высокой средней оценкой книг — учтем только книги с 50 и более оценками.

In [10]:
author_avg_rating = '''
SELECT author,
       ROUND(AVG(rating), 2) as avg_rating
FROM authors AS a
LEFT JOIN books AS b ON b.author_id=a.author_id
LEFT JOIN ratings AS r ON r.book_id=b.book_id
WHERE b.book_id IN (SELECT book_id
                           FROM ratings
                           GROUP BY book_id
                           HAVING COUNT(rating) > 50)
GROUP BY author
ORDER BY AVG(rating) DESC
LIMIT 1;
'''
select(author_avg_rating)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**Вывод:**

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré.

## Задание 5

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
count_avg_reviews = '''
WITH count_avg AS 
   (SELECT username,
           COUNT(review_id) AS count_review
    FROM reviews
    WHERE username IN (SELECT username
                              FROM ratings
                              GROUP BY username
                              HAVING COUNT(rating) > 50)
    GROUP BY username)
    
SELECT ROUND(AVG(count_review), 0)
FROM count_avg;                                  
'''
select(count_avg_reviews)

round
0   24.0

**Вывод:**

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, равно 24.